### ISSUE TITLE
Add Keyboard for other languages

### ISSUE THREAD

**Issue by Moby812** [XITRIX/Moonlight-Switch#203]
Hello!
Is it possible to add keyboard layouts for other languages? For example, Cyrillic?

**Comment by Moby812** [comment]
Thank you, I didn't know that it was already implemented.
And then is it possible to add this layout switcher to the "compact" keyboard? (It is only available in the full-size one)

**Comment by Moby812** [comment]
Is it possible to add the ability (by specifying in the settings which key combination is used) to change the layout on the computer itself? 
And for example, by this specified button, the layout on the computer will switch and the virtual keyboard will switch to a previously specified language (from Russian to English and back. Without opening a submenu).

**Comment by XITRIX** [comment]
You can press "alt shift" on virtual keyboard to switch it on host, like you do it on real keyboard.

Switching keyboards without menu is a nice point, but I'm afraid it will require a lot of time to make it properly, while moonlight was never about "typing", it's about gaming, and this keyboard is more like a nice bonus than a core feature, so probably I will not implement it any time soon (but I like an idea, so it's not impossible)

**Comment by Moby812** [comment]
Thank you very much!
I understand perfectly well that this is a low priority task)
But if I ever get around to it, I would be so happy)

**Comment by moi952** [comment]
> You can press "alt shift" on virtual keyboard to switch it on host, like you do it on real keyboard.
> 
> Switching keyboards without menu is a nice point, but I'm afraid it will require a lot of time to make it properly, while moonlight was never about "typing", it's about gaming, and this keyboard is more like a nice bonus than a core feature, so probably I will not implement it any time soon (but I like an idea, so it's not impossible)

Isn't there a library that allows you to recover ready-made keyboards?
It's true that it would be practical to have my keyboard in French 😜

Maybe using https://github.com/keymanapp/keyman would make it faster?

**Comment by XITRIX** [comment]
Looks like I didn't thought about it 3 years ago :D
Btw thanks for the library, looks nice, but I'm afraid app is already at that point, where it will be quite hard to replace current keyboard with something else. Anyway with current keyboard implementation should be quite easy to create new keyboard layout. I need to find some time to implement layout builder, so anyone could contribute their keyboard to the project

P.S.
Is this how French keyboard looks like? I could try to implement it as an example for layout builder
![image](https://github.com/user-attachments/assets/c530919b-626d-442c-af41-63d320c64860)

**Comment by moi952** [comment]
> Looks like I didn't thought about it 3 years ago :D Btw thanks for the library, looks nice, but I'm afraid app is already at that point, where it will be quite hard to replace current keyboard with something else. Anyway with current keyboard implementation should be quite easy to create new keyboard layout. I need to find some time to implement layout builder, so anyone could contribute their keyboard to the project
> 
> P.S. Is this how French keyboard looks like? I could try to implement it as an example for layout builder ![image](https://private-user-images.githubusercontent.com/9553519/400142535-c530919b-626d-442c-af41-63d320c64860.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYwMDA3NDAsIm5iZiI6MTczNjAwMDQ0MCwicGF0aCI6Ii85NTUzNTE5LzQwMDE0MjUzNS1jNTMwOTE5Yi02MjZkLTQ0MmMtYWY0MS02M2QzMjBjNjQ4NjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMDRUMTQyMDQwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDk2OTA0ZmJiMTFjODhhYzVhYmJlZjhmNTgwZDY2OWEwYTExN2RkMjliZDkyMDBmMGJkN2NmMzBkNGY2YTYwOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.ClSTPb3OdkxSGguH7IFzwlIDNGWL793F2KZIgH0I7II)

It's a French macbook keyboard.
Thanks you're great 😀

This is a french keyboard :

![clavier fr](https://github.com/user-attachments/assets/f29c875f-b5a5-469c-8954-57af45d321bc)

**Comment by XITRIX** [comment]
<img width="1284" alt="image" src="https://github.com/user-attachments/assets/70eabe14-c9d7-455c-ac61-3cd5ee1642be" />
Ok, actually it should be really easy to add new keyboard, just populate new KeyboardLocale to locales array

**Comment by moi952** [comment]
> <img width="1284" alt="image" src="https://github.com/user-attachments/assets/70eabe14-c9d7-455c-ac61-3cd5ee1642be" />
> 
> Ok, actually it should be really easy to add new keyboard, just populating new KeyboardLocale to locales array

We can use chatGpt for that 😁

**Comment by XITRIX** [comment]
UP.

No ... I want to cry, French keyboard literally moves VK_Keycodes around, looks like it's just coincidence that Russian keyboard uses 1:1 the same layout as UK keyboard, while AZERTY keyboards moves buttons around (not just labels of keys)

**Comment by XITRIX** [comment]
<img width="302" alt="image" src="https://github.com/user-attachments/assets/9cacd407-70ca-46fc-9a77-3a90f008ea67" />
Is this layout looks correct?

**Comment by moi952** [comment]
> <img alt="image" width="302" src="https://private-user-images.githubusercontent.com/9553519/400155119-9cacd407-70ca-46fc-9a77-3a90f008ea67.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYwMTI2NDAsIm5iZiI6MTczNjAxMjM0MCwicGF0aCI6Ii85NTUzNTE5LzQwMDE1NTExOS05Y2FjZDQwNy03MGNhLTQ2ZmMtOWE3Ny0zYTkwZjAwOGVhNjcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDEwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMDRUMTczOTAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTM5MWYxZDQ0OWUwN2Q0YzI4MTgwODRiZDk2NjMxYTQ3NDIxMzViNWMxOTc2YjlkODVhMTZlMjg1MjU4YmExYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.7nvqpA8J8sk-F3VlUHw2Mmb79BhvhMOYhv7FTBMNGDA"> Is this layout looks correct?

Yes it is good

**Comment by XITRIX** [comment]
![IMG_9765](https://github.com/user-attachments/assets/ec38fb55-2781-43e5-a9cc-c4ea1439a784)
![IMG_9766](https://github.com/user-attachments/assets/021ed702-a667-4c86-9d0a-b85be2e7cdb9)

Here we go

**Comment by moi952** [comment]
I'll give you a clearer feedback (and written on the right ticket) on the French keyboard

1. The keyboard is always QWERTY when in French. If I press “A”, a “Q” is typed, and so on.
2. The Espace and Suppr keys seem to be translated, but for example, “espace” remains “espace” on the French keyboard when it should be “espace” according to the translations.
3. The Caps Lock key does not lock capital letters. The Shift key does (the two keys seem to work in reverse).

**Comment by XITRIX** [comment]
About Caps Lock, it behave as a regular button, it's only your Host who knows real state of Caps Lock and actually it could get desynchronized with Moonlight, so probably you have correct behaviour of Caps Lock

**Comment by XITRIX** [comment]
@moi952, fixed keyboard layout, check it pls
https://github.com/XITRIX/Moonlight-Switch/actions/runs/12642639448

**Comment by moi952** [comment]
> About CapsLock, there is no way to have any kind of visual feedback to it, because Moonlight has no idea about it's current state on Host (and it's pretty bad idea to try to guess it), also in Windows you can reassign CapsLock key to any other button, so it should behave as regular CapsLock key to perform correctly
> 
> Shift on the other hand could have visual feedback, because I can simulate it's lock position, by sending "key pressed" without "key released" event.
> 
> About Shift's behavior I would agree that it could be improved, but it would make keyboard implementation much more complex, Moonlight was never about typing essays, it's about playing games, and keyboard is just an accessibility tool, that could help you to perform some tasks during game sessions (like enter password, or respond in the chat), so I just don't want to spend a lot of time on it
> 
> P.S. "Delete" button works, it deletes to forward, not backward as "Remove"

Yes yes I completely understand you are not going to spend time on 0.000001% of the use of moonlight switch.
I have sent the information back in case it was simple and quick to implement.

Thank you very much 😀
